Use simulation reduction automatically only in case of custom mediums

dbochkov-flexcompute opened the pull request:
- shrink bounds when performing `grid.discretize_inds` on new region to avoid including possible spurious pixel
- switch simulation reduction in mode solver to be automatic only for simulations with custom medium + display warning if that happens

momchil-flex commented on tidy3d/web/api/mode.py:
```diff
@@ -83,6 +85,21 @@ def run(
+    if reduce_simulation == "auto":
+        sim = mode_solver.simulation
+        contains_custom = any(isinstance(s.medium, AbstractCustomMedium) for s in sim.structures)
+        contains_custom = contains_custom or isinstance(sim.medium, AbstractCustomMedium)
```
We do something like that in a bunch of places but yeah it seems like there's still no in-built way to get "all media, including the simulation one" and "all structures, including the simulation as a structure"...

dbochkov-flexcompute replied:
Well, I guess I could've just used `Simulation.mediums`

momchil-flex replied:
Ah yeah so there is something like that!

momchil-flex commented on tidy3d/web/api/mode.py:
```diff
@@ -83,6 +85,21 @@ def run(
+                " Setting `reduce_simulation=True` will force simulation reduction in all cases and"
+                " silence this warning."
+            )
+
```
I think it looks good, probably we should update the notebooks then to avoid the warning?

dbochkov-flexcompute replied:
was there any other besides BullseyeCavityPSO?

momchil-flex replied:
BullseyeCavityPSO will not have a warning anymore as it doesn't use custom medium, I was thinking of the heat notebooks actually.

momchil-flex commented on tidy3d/plugins/mode/mode_solver.py:
```diff
@@ -973,7 +973,8 @@ def reduced_simulation_copy(self):
         """
 
-        # we preserve extracells along the normal direction to ensure
+        # we preserve extracells along the normal direction to ensure there is enough data for
```
extra cells

dbochkov-flexcompute commented:
addressed the comments, also updated heat notebooks https://github.com/flexcompute/tidy3d-notebooks/pull/24

note, there is a change in web api container.py, I noticed that it was causing spurious errors due to pydantic trying to initialize the field `Batch.simulations` with every possible simulation class
